[Wong Zhi, Chester] iP

Chesterwongz opened the pull request:
### Chester's Duke

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Chester's Duke frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is,


1. download it from [here](https://github.com/Chesterwongz/ip/releases/tag/v0.1).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉

And it is **FREE**!

Features:

- [x]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [x]  Reminders (coming soon)

------------------------------------------------------------------------------------------

If you're a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(Duke.class, args);
    }
}
```

Commit:
Add Gradle support

Chesterwongz committed:
Level-1. Duke can now greet, echo and say goodbye, with some formatting

Signed-off-by: Chesterwongz <psyforce99@gmail.com>

Chesterwongz committed:
Level-2 done. Changed if else to switch when checking input and removed checkExit function

Chesterwongz committed:
Level-3

Tasks can now be marked as done.  Only works when task to be marked is a valid number. This will be address in Level-5 error handling

Chesterwongz committed:
Level-4

ToDos, Events, Deadlines done with switches.

Chesterwongz committed:
A-TextUiTesting

Automated UI regression testing now possible

Chesterwongz committed:
exception handling

Most, if not all possible errors are being handled in the current version.

Chesterwongz committed:
Level-6

Added delete function and made code neater. Changed Task array to use ArrayList.

Chesterwongz committed:
fixed java docs missing docs

Chesterwongz committed:
Save tasks to hard disk and load on start up

Our task list is lost when Duke is exited.

Storing the task list somewhere improves Duke's usability.

Let's save our tasks in a file which we will load whenever Duke restarts.

Chesterwongz committed:
Automatically save tasks whenever task list changes

The saved task list is not updated if Duke crashes.

Updating the task list automatically on change ensures saved tasks will be up-to-date even if Duke terminates prematurely.

Let's write our tasks to the file whenever the user adds one.

Chesterwongz committed:
Add date and time formating

Chesterwongz committed:
Add more OOP

Chesterwongz committed:
Organize code into packages

Chesterwongz committed:
Merge branch 'master' into branch-A-Packages

Chesterwongz committed:
Make actual and expected text match in automated testing

Chesterwongz committed:
Add JUnit unit testing

Chesterwongz committed:
Add JavaDoc Comments

Chesterwongz committed:
Give users a way to find a task by searching for a keyword

Chesterwongz committed:
Delete tasks.txt

Chesterwongz committed:
Merge branch 'branch-Level-9'

Chesterwongz committed:
Follow coding standards

Chesterwongz committed:
Add JavaDocs to public constructors

Chesterwongz committed:
Merge branch 'branch-A-JavaDoc'

Chesterwongz committed:
Update function names to be more relevant

Chesterwongz committed:
Create Jar for distribution

Chesterwongz committed:
Add class header comments

Chesterwongz committed:
Merge remote-tracking branch 'upstream/add-gradle-support' into branch-A-Gradle

Chesterwongz committed:
Automate project builds using Gradle

BenjaminTan99 commented on src/main/duke/data/TaskList.java:
```diff
@@ -0,0 +1,61 @@
+        taskList = new ArrayList<>();
+    }
+
+    public Task get(int index) {
```
You should specify in the method name what you are getting - in this case getTask

BenjaminTan99 commented on src/main/duke/data/TaskList.java:
```diff
@@ -0,0 +1,61 @@
+        return taskList.remove(index);
+    }
+
+    public int size() {
```
ArrayList itself already has a size() method - seems unnecessary but it's code preference.

BenjaminTan99 commented on src/main/duke/data/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+ * This class abstracts a deadline task.
+ */
+public class Deadline extends Task {
+    protected final LocalDateTime by;
```
Final variables/constants should be in full capital letters according to the Java Coding Standard.

BenjaminTan99 commented on src/main/duke/data/task/Deadline.java:
```diff
@@ -0,0 +1,42 @@
+     */
+    public Deadline(String description, LocalDateTime by) {
+        super(description);
+        this.by = by;
```
Amend 'by' as mentioned earlier.

BenjaminTan99 commented on src/main/duke/data/task/Event.java:
```diff
@@ -0,0 +1,42 @@
+ * This class abstracts an Event task.
+ */
+public class Event extends Task {
+    protected final LocalDateTime at;
```
Final variables/constants should be in full capital letters.

BenjaminTan99 commented on src/main/duke/data/task/Event.java:
```diff
@@ -0,0 +1,42 @@
+    /**
+     * Returns the String representation of an Event object.
+     *
+     * @return THe String representation of an Event object.
```
A bit of typo here 😄

BenjaminTan99 commented on src/main/duke/data/task/Task.java:
```diff
@@ -0,0 +1,48 @@
+ * This class abstracts a task.
+ */
+public abstract class Task {
+    protected final String description;
```
Final variables/constants should be in full capital letters.

BenjaminTan99 commented on src/main/duke/data/task/Task.java:
```diff
@@ -0,0 +1,48 @@
+    protected boolean isDone;
+
+    protected Task(String description) {
+        this.description = description;
```
As mentioned above for final variables/constants.

BenjaminTan99 commented on src/main/duke/storage/Storage.java:
```diff
@@ -0,0 +1,134 @@
+                storageList.add(eventTask);
+            }
+        } catch (DateTimeParseException e) {
+            // Invalid data, do nothing.
```
You might want to print a message indicating the correct/ideal format for date time instead of leaving it blank so that the user can use your program better. You can even include a reminder message regarding this upon startup!

BenjaminTan99 commented on src/main/duke/storage/Storage.java:
```diff
@@ -0,0 +1,134 @@
+                storageList.add(deadlineTask);
+            }
+        } catch (DateTimeParseException e) {
+            // Invalid data, do nothing.
```
Similar comments for readEvent().

BenjaminTan99 commented on src/test/duke/DeadlineCommandTest.java:
```diff
@@ -0,0 +1,51 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class DeadlineCommandTest {
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
```
As mentioned earlier, final variables/constants should be in full capital letters.

BenjaminTan99 commented on src/test/duke/EventCommandTest.java:
```diff
@@ -0,0 +1,51 @@
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class EventCommandTest {
+    private final ByteArrayOutputStream outContent = new ByteArrayOutputStream();
```
As mentioned earlier, final variables/constants should be in full capital letters.

BenjaminTan99 reviewed (commented):
Besides some minor changes to be made, the code overall looks clean and LGTM! Can tell you took the time to JavaDocs properly! 😃

Justinhoejj commented on src/main/duke/command/DeadlineCommand.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class DeadlineCommand extends Command {
+    public static final String COMMAND_WORD = "deadline";
+    private final Deadline newTask;
```
Consider capitalising newtask since it is final.

Justinhoejj replied:
Variable newTask implies that there is an old task. Would it be better to name it as task or deadlineTask?

Justinhoejj commented on src/main/duke/command/DeleteCommand.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        if (0 <= taskNum && taskNum < tasks.size()) {
```
Consider flipping 0<= taskNum to tasknum >= 0 for more intuitive reading.

Justinhoejj commented on src/main/duke/command/DoneCommand.java:
```diff
@@ -0,0 +1,43 @@
+     */
+    @Override
+    public void execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        if (0 <= taskNum && taskNum < tasks.size()) {
```
Consider flipping 0<= taskNum to tasknum >= 0 for more intuitive reading.

Justinhoejj commented on src/main/duke/command/ToDoCommand.java:
```diff
@@ -0,0 +1,39 @@
+ */
+public class ToDoCommand extends Command {
+    public static final String COMMAND_WORD = "todo";
+    protected final ToDo newTask;
```
Any reason why newTask here is protected instead of private?

Justinhoejj commented on src/test/duke/ToDoTest.java:
```diff
@@ -0,0 +1,22 @@
+
+public class ToDoTest {
+    @Test
+    public void testConvertToDoToData() {
```
Review all test methods to follow naming format of `featureUnderTest_testScenario_expectedBehavior()` where applicable.

Justinhoejj commented on src/main/duke/ui/Ui.java:
```diff
@@ -0,0 +1,126 @@
+ * @author Chesterwongz
+ */
+public class Ui {
+    private static final String line = "__________________________________________________________";
```
Capitalise variable since it is final.

Justinhoejj reviewed (commented):
Overall well written, easy to read. Do note the convention for naming of test methods. Otherwise great job with the project! 😁

Chesterwongz committed:
Add GUI for duke

Chesterwongz committed:
Add checkstyle to project

Chesterwongz committed:
Fix test cases

Chesterwongz committed:
Make DialogBox handle text overflow

DialogBox cuts text if if overflows. It cannot display more than 5 lines.

Allowing DialogBox to expand with the text will ensure the entire task list is viewable.

lets make DialogBox wrap the text in a flexible way.

Chesterwongz committed:
Add assertions

Chesterwongz committed:
Merge pull request #1 from Chesterwongz/branch-A-Assertions

Add assertions

Chesterwongz committed:
Follow Code Quality Standards

Chesterwongz committed:
Merge pull request #2 from Chesterwongz/branch-A-CodeQuality

Follow Code Quality Standards

Chesterwongz committed:
Added C-MassOps extension feature

Chesterwongz committed:
Add representative screenshot

Chesterwongz committed:
Set theme jekyll-theme-cayman

Chesterwongz committed:
A-UserGuide

Chesterwongz committed:
Ensure method/class header comments follow the format specified in the coding standard

Chesterwongz committed:
Fix parsing bugs
